File: main/java/androidtown/org/fragment_grade.java
package androidtown.org;

import android.graphics.Color;
import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.mikephil.charting.animation.Easing;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.formatter.ValueFormatter;
import com.github.mikephil.charting.utils.ColorTemplate;

import java.text.DecimalFormat;
import java.util.ArrayList;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link fragment_grade#newInstance} factory method to
 * create an instance of this fragment.
 */
public class fragment_grade extends Fragment {
    PieChart pie;
    private ArrayList<PieEntry>entries;
    private PieDataSet set;
    private PieData data;

    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;

    public fragment_grade() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment fragment_grade.
     */
    // TODO: Rename and change types and number of parameters
    public static fragment_grade newInstance(String param1, String param2) {
        fragment_grade fragment = new fragment_grade();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_grade, container, false);

        pie=view.findViewById(R.id.pie_chart);
        setData();
        setUpData();


        return view;
    }
    public void setData(){
        entries=new ArrayList<>();
        entries.add(new PieEntry(34,"전공필수"));
        entries.add(new PieEntry(19,"전공선택"));
        entries.add(new PieEntry(18,"교양선택"));
        entries.add(new PieEntry(13,"교양필수"));
        entries.add(new PieEntry(3,"교양계열"));

        set=new PieDataSet(entries,"");
        set.setColors(ColorTemplate.JOYFUL_COLORS);
        set.setSelectionShift(3f);
        //set.setXValuePosition(PieDataSet.ValuePosition.OUTSIDE_SLICE);
        //set.setYValuePosition(PieDataSet.ValuePosition.OUTSIDE_SLICE);
        //set.setUsingSliceColorAsValueLineColor(true);

        data=new PieData(set);
        data.setValueTextSize(12f);
        data.setValueFormatter(new ValueFormatter() {
            @Override
            public String getFormattedValue(float value) {
                return String.format("%d", (int) value);
            }
        });
    }
    public void setUpData(){
        pie.setData(data);
        pie.setEntryLabelColor(Color.BLACK);
        pie.getDescription().setEnabled(false);
        pie.setUsePercentValues(true);
        pie.animateY(2000, Easing.EaseInOutQuad);

        Legend legend=pie.getLegend();
        legend.setVerticalAlignment(Legend.LegendVerticalAlignment.TOP);
        legend.setHorizontalAlignment(Legend.LegendHorizontalAlignment.CENTER);

        pie.setCenterText("취득학점\n"+new DecimalFormat("###").format(getSum()));
        pie.setCenterTextSize(30f);

        pie.invalidate();
    }
    public int getSum(){
        int sum=0;
        for(int i=0;i<entries.size();i++){
            sum+=entries.get(i).getValue();
        }
        return sum;
    }
}

File: app/src/main/java/androidtown/org/fragments/FragmentGrade.kt
package androidtown.org.fragments

import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.Handler
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import androidtown.org.R
import androidtown.org.listener.GradeButtonListener
import androidx.core.view.allViews
import androidx.fragment.app.Fragment
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.utils.ColorTemplate
import java.text.DecimalFormat

class FragmentGrade(private val listener: GradeButtonListener) : Fragment() {
    private lateinit var pie: PieChart
    private lateinit var entries: ArrayList<PieEntry>
    private lateinit var set: PieDataSet
    private lateinit var data: PieData
    private lateinit var gradeTable: TableLayout

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_grade, container, false)

        pie = view.findViewById(R.id.pie_chart)
        gradeTable = view.findViewById(R.id.gradeTable)
        Handler().postDelayed({
            setData()
            setUpData()
            updateGradeTable()
        }, 500)

        return view
    }

    private fun updateGradeTable() {
        listener.gradeMap.entries.forEachIndexed { index, (k, v) ->
            val row = TableRow(context).apply {
                id = k.toInt() + index
                background = Drawable.createFromPath("@drawable/border")
            }

            val subject = TextView(context).apply {
                text = v.subjectName
                width = 200
                gravity = Gravity.CENTER
                textSize = 17f
            }

            val isuName = TextView(context).apply {
                text = v.isuName
                width = 85
                gravity = Gravity.CENTER
                textSize = 17f
            }

            val credit = TextView(context).apply {
                text = v.credit
                width = 85
                gravity = Gravity.CENTER
                textSize = 17f
            }

            val grade = TextView(context).apply {
                text = v.rank
                width = 85
                gravity = Gravity.CENTER
                textSize = 17f
            }

            row.addView(subject)
            row.addView(isuName)
            row.addView(credit)
            row.addView(grade)

            gradeTable.addView(row)
        }
    }

    private fun setData() {
        val graduate = listener.graduate
        entries = ArrayList()
        entries.add(PieEntry(graduate.majbas_credit.toFloat(), "전공필수"))
        entries.add(PieEntry(graduate.majsel_credit.toFloat(), "전공선택"))
        entries.add(PieEntry(graduate.culess_credit.toFloat(), "교양필수"))
        entries.add(PieEntry(graduate.comb_cul_tot_credit.toFloat(), "취득영역 소계"))
        entries.add(PieEntry(graduate.culbranch_credit.toFloat(), "계열 교양"))

        set = PieDataSet(entries, "")
        set.setColors(*ColorTemplate.JOYFUL_COLORS)
        set.selectionShift = 3f

        data = PieData(set)
        data.setValueTextSize(12f)
        data.setValueFormatter(object : ValueFormatter() {
            override fun getFormattedValue(value: Float): String {
                return String.format("%d", value.toInt())
            }
        })
    }

    private fun setUpData() {
        pie.data = data
        pie.setEntryLabelColor(Color.BLACK)
        pie.description.isEnabled = false
        //pie.setUsePercentValues(true);
        pie.animateY(2000, Easing.EaseInOutQuad)

        val legend = pie.legend
        legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
        legend.horizontalAlignment = Legend.LegendHorizontalAlignment.CENTER

        pie.centerText = """
             취득학점
             ${DecimalFormat("###").format(sum.toLong())}
             """.trimIndent()
        pie.setCenterTextSize(30f)

        pie.invalidate()
    }

    private val sum: Int
        get() {
            var sum = 0
            for (i in entries.indices) {
                sum = (sum + entries[i].value).toInt()
            }
            return sum
        }
}